katcp/test/test_server.py

```diff
@@ -462,6 +541,127 @@ def test_request_timeout_hint(self):
+        client = self.set_up_client_for_bulk_sensor_sampling()
+
+        req = mock_req(
+            "sensor-sampling", "an.int,a.float", "event", client_conn=client
+        )
+        self.server.request_sensor_sampling(req, req.msg).result(timeout=1)
+        inform = req.client_connection.inform
+        inform.assert_wait_call_count(count=2)
+        inform_messages = [call.args[0] for call in inform.call_args_list]
+        self._assert_msgs_equal(
+            inform_messages, (
+                r"#sensor-status 12345.000000 1 an.int nominal 3",
+                r"#sensor-status 12345.000000 1 a.float nominal 12.0",
+            )
+        )
+
+    def test_bulk_sensor_sampling_fails_for_invalid_sensor_name(self):
+        client = self.set_up_client_for_bulk_sensor_sampling()
+
+        req = mock_req(
+            "sensor-sampling", "an.int,an.invalid", "period", 2.5, client_conn=client
+        )
+        with self.assertRaises(FailReply):
+            self.server.request_sensor_sampling(req, req.msg).result()
+
+    def test_bulk_sensor_sampling_fails_for_empty_sensor_name(self):
+        client = self.set_up_client_for_bulk_sensor_sampling()
+
+        req = mock_req(
+            "sensor-sampling", "", "period", 2.5, client_conn=client
+        )
+        with self.assertRaises(FailReply):
+            self.server.request_sensor_sampling(req, req.msg).result()
+
+    def test_bulk_sensor_sampling_fails_for_invalid_strategy(self):
+        client = self.set_up_client_for_bulk_sensor_sampling()
+
+        req = mock_req(
+            "sensor-sampling", "an.int,a.float", "invalid-strategy", 2.5,
+            client_conn=client
+        )
+        with self.assertRaises(FailReply):
+            self.server.request_sensor_sampling(req, req.msg).result()
+
+    def test_bulk_sensor_sampling_fails_for_valid_strategy_invalid_params(self):
+        client = self.set_up_client_for_bulk_sensor_sampling()
+
+        req = mock_req(
+            "sensor-sampling", "an.int,a.float", "period", 2.5, 1.5, client_conn=client
+        )
+        with self.assertRaises(FailReply):
+            self.server.request_sensor_sampling(req, req.msg).result()
+
+    def test_bulk_sensor_sampling_fails_for_invalid_strategy_invalid_params(self):
+        client = self.set_up_client_for_bulk_sensor_sampling()
+
+        req = mock_req(
+            "sensor-sampling", "an.int,a.float", "invalid-strategy", 2.5, 1,
+            client_conn=client
+        )
+        with self.assertRaises(FailReply):
+            self.server.request_sensor_sampling(req, req.msg).result()
+
+    def test_bulk_sensor_sampling_fails_for_no_strategy(self):
+        client = self.set_up_client_for_bulk_sensor_sampling()
+
+        req = mock_req("sensor-sampling", "an.int,a.float", client_conn=client)
+        with self.assertRaises(FailReply):
+            self.server.request_sensor_sampling(req, req.msg).result()
+
```

Comment:
One more test has come to mind.  The spec says that the implementation must be atomic.  I.e, if we can't set the strategies on all the sensors, then there must be no change to any of them.  Something like: 
`test_bulk_sensor_sampling_all_strategies_unchanged_for_failure`.